httpx/middleware/auth.py

```diff
@@ -0,0 +1,249 @@
+        "MD5": hashlib.md5,
+        "MD5-SESS": hashlib.md5,
+        "SHA": hashlib.sha1,
+        "SHA-SESS": hashlib.sha1,
+        "SHA-256": hashlib.sha256,
+        "SHA-256-SESS": hashlib.sha256,
+        "SHA-512": hashlib.sha512,
+        "SHA-512-SESS": hashlib.sha512,
+    }
+
+    def __init__(
+        self, username: bytes, password: bytes, per_nonce_count: typing.Dict[bytes, int]
+    ) -> None:
+        self.username = username
+        self.password = password
+        self.per_nonce_count = per_nonce_count
+        self.num_401_responses = 0
+
+    async def __call__(
+        self, request: AsyncRequest, get_response: typing.Callable
+    ) -> AsyncResponse:
+        response = await get_response(request)
+        if not (
+            StatusCode.is_client_error(response.status_code)
+            and "www-authenticate" in response.headers
+        ):
+            self._num_401_responses = 0
+            return response
+
+        header = response.headers["www-authenticate"]
+        try:
+            challenge = DigestAuthChallenge.from_header(header)
+        except ValueError:
+            raise ProtocolError("Malformed Digest authentication header")
+
+        if self._previous_auth_failed(challenge):
+            return response
+
+        request.headers["Authorization"] = self._build_auth_header(request, challenge)
+        return await self(request, get_response)
+
+    def _previous_auth_failed(self, challenge: "DigestAuthChallenge") -> bool:
+        """Returns whether the previous auth failed.
+
+        This is fairly subtle as the server may return a 401 with the *same* nonce that
+        that of our first attempt. If it is different, however, we know the server has
+        rejected our credentials.
+        """
+        self.num_401_responses += 1
+
+        return (
+            challenge.nonce not in self.per_nonce_count and self.num_401_responses > 1
+        )
+
+    def _build_auth_header(
+        self, request: AsyncRequest, challenge: "DigestAuthChallenge"
+    ) -> str:
+        hash_func = self.ALGORITHM_TO_HASH_FUNCTION[challenge.algorithm]
+
+        def digest(data: bytes) -> bytes:
+            return hash_func(data).hexdigest().encode()
+
+        A1 = b":".join((self.username, challenge.realm, self.password))
+
+        path = request.url.full_path.encode("utf-8")
+        A2 = b":".join((request.method.encode(), path))
+        # TODO: implement auth-int
+        HA2 = digest(A2)
+
+        nonce_count, nc_value = self._get_nonce_count(challenge.nonce)
+        cnonce = self._get_client_nonce(nonce_count, challenge.nonce)
+
+        HA1 = digest(A1)
+        if challenge.algorithm.lower().endswith("-sess"):
+            HA1 = digest(b":".join((HA1, challenge.nonce, cnonce)))
+
+        qop = challenge.qop
+        if not qop:
+            digest_data = [HA1, challenge.nonce, HA2]
+        elif qop == b"auth" or b"auth" in qop.split(b",") or b"auth" in qop.split(b" "):
```

Comment:
So my comment about being able to handle spaces was we have to be able to handle `qop` values like the one in [RFC 7616 Section 3.9.1](https://tools.ietf.org/html/rfc7616#section-3.9.1) so we should pull all values via something like:

`qop = re.split(b", ?", challenge.qop)`

and then check for inclusiveness (`"auth" in qop`). We should only raise the `NotImplementedError` if `auth` isn't in the list.